### ISSUE TITLE
Reduce dome movements during tracking

### ISSUE THREAD

**Issue by mworion** [mworion/MountWizzard4#159]
During tracking the amount of dome-slews is quite large. Every 3-5 seconds the dome is instructed to follow the scope. I have a dome-slit-width programmed in at 0.81 metres, but the routine does not seem to make use of that knowledge.
So the functionality works perfect, but could do with some improvement regarding the number of corrections for the slit-width. I do realise that it is not quite easy to do the required maths, but it would be great if dome slewing could be done as follows:

start with the slit centred over the scope;
once tracking, move the dome when about 10-20% near the slit's railing edge;
then move the dome towards the leading edge (again some 10-20% clear of it);
wait until the trailing edge is reached again.
As the satellite trajectories are curved in respect to the dome slit, the above steps require knowledge of the complete trajectory. Calculating the optimum slew-angle has to be done for several points along the trajectory to ensure the dome is not slewed too far off.

The reason I would appreciate the number of dome-slews to be minimised is that each dome slew has its effect on the pier (despite the vibrations going down one floor to reach the base of the pier), which is especially noticeable at f/40 with the C11.

Once again thanks for this improvement! Hopefully you are able to implement above suggestion viz. the reduction of slew commands as well.

Nicolàs

**Comment by InFINNity-Deck** [comment]
Hi Michel,

for completeness sake: above comments came from me from [another issue](https://github.com/mworion/MountWizzard4/issues/139#issuecomment-881893045). To that you replied:

```
... For the Dome movements (as I don't have one) I need some more context, how dome is functioning. Currently I send always if I correct the the Azimuth also a new dome position. There is right now no relation to the slit width. I recognized that some does controllers (like Aries one) have a configured hysteresis where it does not react at all (e.g. 2 deg). So any chance within this value, the dome does not move. I struggled with that as the dome Dows not reach the traget position (and I had the dome slewing state present). In addition if I take care of the slit with (which is doable), the hysteresis parts are summing up. Problem is that hysteresis settings are normally not readable from the dome controllers nor being able to switch them on or off. Plus as you already showed, depending of the position the resulting dome slit with is due to change if the scope is not located in the center line (which is always the case in side-by-side setups).

If you have some thoughts about how to address this, please let me know (especially how you as a dome user interact with these unlinked controls)
Michel
```

Indeed hysteresis is present in most, if not all, dome set-ups, and so is it in mine (I need to check, but I think it is about 2 or 3 degrees). That is the reason why I wrote that a margin should be kept from the side of the slit to avoid that the dome is slewing too far. This can be achieved in two ways:

- either the user enters a narrower slit (but that may be confusing);
- or MW4 will get an additional field in which a tolerance can be supplied in metres, degrees or percentage, depending what is easiest to implement (but that is more work for you).

I will give this a further thought as well!

Nicolàs

**Comment by mworion** [comment]
Hi Nicolas,
I had thoughts in the same direction. My plan is described in the documents here (last entry):
[https://mountwizzard4.readthedocs.io/en/latest/config/dome.html](url)
Please let me know, how you think about it. I also struggling with the naming of the parameters.
Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

about nomenclature, I think it is good to follow a major manufacturer, see one of their [plan-views](https://observa-dome.com/wp-content/uploads/2014/11/25V101.pdf) and [front-and-side elevation](https://observa-dome.com/wp-content/uploads/2014/11/25V100.pdf):

- de opening of a dome is called the slit;
- the available aperture (sideways) of the slit is called the "slitframe opening" or "clear opening";
- the slit is closed with shutters;
- the "shutter zenith distance" is also called "clearance from zenith".

I appreciate you wrote this ["Configuring dome geometry"-page](https://mountwizzard4.readthedocs.io/en/latest/config/dome.html), that clarifies a lot. The section about the vertical offset reads:

> For the vertical displacement you have to take car about different definitions of how to measure this value. A common way many ASCOM drivers do is measuring the distance between the dome center and the GEM point of the mount. The GEM point of the mount is the intersection of the RA and DEC axis. Ideally you will have that point fitting to the dome center. In the 10micron handbook, the measurement is done between the dome center and the base of the mount. This value is mostly negative, because the mount base level is on lower height than the dome center.

I think it would be good to add to it that the definition used by MW4 is: "In MW4 the vertical offset is positive if the GEM point of the mount is above the dome centre."

The section about the lateral offset reads (I highlighted a small typo):

> If you have **more** than one telescope mounted, you will experience a lateral displacement of the OTA you would like to use for reference. The measurement is taken from base plate of the mount to the OTA axis. Displacement to the right (east) is treated positive.

Perhaps it would be better to rephrase the last sentence as "With the scopes pointing towards Polaris and while looking to the north a displacement to the right (east) is treated positive."
What directly comes to mind is how this is done on the southern hemisphere, so perhaps this needs another explanatory line.

Nicolàs

**Comment by mworion** [comment]
Nicolas,
v2.0.0b26 should address this point. I refined the naming and implemented the new following stategy.
Michel

**Comment by InFINNity-Deck** [comment]
Great, thanks! :-)

Nicolàs

**Comment by InFINNity-Deck** [comment]
Hi Michel,

I tested the new feature (Enable Dynamic Following), but there appears to be some kind of bug in the code. Here is the log:

[mw4-2021-07-21.log](https://github.com/mworion/MountWizzard4/files/6855677/mw4-2021-07-21.log)

My observations:
- dome is slewed only once, then MW4 stops;
- setting focal length and aperture did not solve this issue;
- taking over the focal length from driver does not seem to work;
- tooltips on focal length, aperture and "Automatic takeover from driver" are incorrect.

The few slews I saw made me wonder if aperture (or rather the aperture radius) is taken into account in the routine? After all we want to avoid that even the smallest part of the objective is facing the side of the dome slit. So if the tolerance is set to 0%, the dome should still move when the side of the objective is not within the slit area.

I also wonder that if no aperture is specified under Telescope Information Readout that the user should get a warning when using this dynamic following.

With dynamic following disabled all works as expected, had a very good run last night again! :-)

Nicolàs

**Comment by mworion** [comment]
Hi Nicolas,

corrected the Tooltips.
Focal length and aperture are not taken into account for the dome slewing algorithm as it does not make sense - or better you set the opening hysteresis for adjusting the "border" when the dome will slew again.

So out of the description: 
Opening hysteresis describes the distance from center "beam" of telescope to dome opening borders.

I you first run with dynamic following there is hardly any communication to the mount and dome. So either Trace was not enabled or there was a communication problem.

Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

that is correct, trace mode was off during the first run. It must have been on during the second, and maybe off again during the third (the third was after I had shut down the observatory and on turning it on it reverted back to normal as I do not have trace  on by default).

Ok, so for the hysteresis one should add semi-aperture and hysteresis together. In my case that will be 7.5cm for the scope and 8cm for uncertainty of dome slew (tan(3°)x150cm), so together 15.5cm, rounded up this makes 16cm.

Please let me know if you want me to produce a new log-file using 2.0.0b26.

Nicolàs

**Comment by mworion** [comment]
Nicolas,
I did some improvements mentioned in other threads. So v2.0.0b27 is the new one. If you don't mind, please do a test with it.
Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

tried dynamic following again in Trace mode (log-file attached, see below). Did three runs:

- Dynamic following enabled
- Dynamic following disabled
- Dynamic following enabled

The second test was done to see if it would continue working, which it did. The Dynamic Following seems to stall the model building process as soon as no dome-move is required. So the mount slews to the target and when no dome move is required, no image is taken, and MW4 halts.

A few other observations:
- the new internal (beta) loader works like a charm! Massive improvement! :-)
- the hysteresis setting does not allow values above 0.15m (for my C11 I would need 0.23m). I think the upper limit can be coupled to the slit-width as more than 50% of the slit-width as tolerance makes no sense;
- there is a typo in the tooltip for focal length ("fical");
- there are no tooltips for the dynamic following section.

That last point I noticed when I was wondering what "Overshoot" option of dynamic following would do. Can you explain please?

Nicolàs

[mw4-2021-07-22.zip](https://github.com/mworion/MountWizzard4/files/6865146/mw4-2021-07-22.zip)

**Comment by mworion** [comment]
Nicolas,
Your feedback helped me to understand. I think I know what to do for correcting the issue.
To the overshoot: if dynamic works, MW4 will not slew the dome until crossing the hysteresis and slews the dome when crossed to the center (which is the known next position). Overshoot moves the dome even further (0-100%) of the last azimuth increase to reduce dome movements. As I do not know the next azimuth ist just "more". You have to experiment.
Michel
But I have to build a next beta to correct the issues you mentioned.

**Comment by InFINNity-Deck** [comment]
Hi Michel,

glad that my info helped! The overshoot function is very nice, but of course it will require an east-west sorting of the points (which I do). I will play with it once the new beta version is out.

thanks again,
Nicolàs

**Comment by mworion** [comment]
Nicolas,
made a short video, how it works now: https://youtu.be/ZmDz-rtvFzc
Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

thanks! I have tested 2.0.0b29 and indeed the amount and length of dome-slews have decreased, so that is great. Especially those very short jerky dome-slews are gone now. The overshoot I am not sure if it is useful as it is implemented right now. I have for instance four model points above each other in azimuth 90 degrees. They have altitudes of 30, 60, 75 and 85 degrees. I sort the points east-west as a result of which the bottom one is done first, then the one above it, etc. For the dome this means that the exact dome position is clockwise for each consecutive point, even though all the points are sorted in a counter-clockwise order. So when using the overshoot function, the dome slews too far counter-clockwise for all four points to fit, as a result of which the dome is sent clockwise for the last two points and that is again done too far. Perhaps a solution would be if I altered the sequence for those points at 90 degrees by giving them slightly different azimuth values (i.e. 90.004 degrees for the top one, 90.003 degrees for the one below that etc.). In that way sorting the points would force MW4 to take the top one first and the overshoot will be effective again.

I am going to think a bit more about this...

Meanwhile many thanks again for these improvements! :-)
Nicolàs

**Comment by mworion** [comment]
Hi Nicolas,
overshoot is still a challenge. I revised it now just to switch it on and off. If you don't mind, please post the buildpoints file and the dome parameters you have chosen. I would like to check where it causes problems.
Still I made v30 which should not slew back and forth with overshoot. Still not working in satellite tracking.
Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

please find attached the buildpoints file.

[manual_grid74pts.zip](https://github.com/mworion/MountWizzard4/files/6874476/manual_grid74pts.zip)

For satellite tracking stuff can me much more complicated as they can fly in any direction and at any side of the observatory. So overshoot has to be able to cope with objects that fly clockwise and counter-clockwise in respect to the observatory.

Nicolàs

**Comment by mworion** [comment]
Hi Nicolas,
So I think finally I made it also with overshoot. I will release v31 this evening. Only point with you modeling setup: For overshoot I have to check in which direction the mount will change it's azimuth in order to "overshoot" or better to "oversleep" the dome. If you have sorted the points in a way that they are not sorted in a single azimuth directions (which should be the case), the dome will slew still more than needed. 
Michel

**Comment by InFINNity-Deck** [comment]
Hi Michel,

great to hear that you found a solution. Although the points in my file are not sorted east-west, I do have the east-west sorting option enabled in MW4, so I presume that should suffice? If not, I can also sort the points in the file if required.

Nicolàs

**Comment by ACMBakx** [comment]
With dynamic following enabled, my Scopedome does NOT follow at all.
Just an observation.

**Comment by mworion** [comment]
Arie, something is wired. Since our work on geometry I have the scopedome driver installed and I did a test again it. Strange. As the driver works with really slow slew speed when in simulation mode (it recognizes that no real dome is present) it takes 2-3 minutes for 90 degrees azimuth change. This makes testing reality hard. Do you have any Clou how to rise slew speed in that mode ?
Michel

Von meinem iPad gesendet

**Comment by mworion** [comment]
Arie,
checked this now with Scopedome driver, the opening clearing and settings should be ok, dome should not move always, but with v2.0.0b39 I could see the scope dome driver moving for the points with dynamic following enabled
Michel

**Comment by ACMBakx** [comment]
Dome still not moving when option is selected.
Image windows shows however that instructions are given, but dome does not move.
With option not selected , dome slews ok.
See log. Did a few slews and took a few images for platesolving.
[mw4-2021-08-13.log](https://github.com/mworion/MountWizzard4/files/6985093/mw4-2021-08-13.log)

**Comment by mworion** [comment]
Hi Arie,
thanks for the testing. I revised some things, added some logging.
It should be solved now.
For clarification on my side it would help, if you could provide your geometry values entered on the dome page (snapshot)
Michel
ps: uploading v40 later this afternoon

**Comment by ACMBakx** [comment]
The Lacerta focuser again refuses to connect.
We had it fixed, but it refuses again. 👎 

Attached a image from MW4 settings
![DomeGeometry](https://user-images.githubusercontent.com/59890758/129456259-4144104d-bebd-4102-bf17-fa9fe69be9b2.JPG)

**Comment by mworion** [comment]
Arie,
Ok this confirms my guess about dynamic following. My default values (and the lower box dome shutter shows them) don't leave any chance for the dynamic following. So this is fixed. To use it best, you need to set them according to your setup. If you have the 2m Scope dome, the shutter width (clear opening) is 0.59 m, the clearance from zenith (distance how much the dome more opens over zenith) is 0.1m. The opening hysteresis should somehow reflect the with of you OTA, so that you don't look to the edge of your dome partly, So it has to be more than half of the aperture.

Could you post the actual log for v39? Would finally check that.

Nothing changed on the focuser side. I check this and come back.
Michel

**Comment by ACMBakx** [comment]
[mw4-2021-08-14.log](https://github.com/mworion/MountWizzard4/files/6987096/mw4-2021-08-14.log)

Sometimes the focuser does connect. Mostly not.
A while ago you've been adjusting some timer, to give it time to be detected.
Probably a Lacerta thing. No issue for me. I usually run an autofocus in NINA first.

**Comment by ACMBakx** [comment]
[mw4-2021-08-14.log](https://github.com/mworion/MountWizzard4/files/6987107/mw4-2021-08-14.log)
Tried to connect the focuser in this one. No go.
Mount was powered off.
Then accidentally started MW4 again . . . .

**Comment by mworion** [comment]
Arie,
I downloaded the latest Lacerta focuser driver. I see comparable results. As I do not own a lacerate focuser, I don't know if that is related to the "open end". But I try to revise the behavior. It might be helpful to other ascom drivers as well.
Michel 

Thanks for the new log. There is somehow a race condition. I have to sort this really out.

**Comment by ACMBakx** [comment]
OK.
FYI I will be using the Pegasus UPB V2 as focus controller when I get that back from repair. (damaged USB port)
The ascom driver from Lacerta is not very good (to put it politely). Most of the time (99.9999%) I cannot access the controller in the ascom driver, to change settings like backlash e.g.  Just about all settings I have to do on the Lacerta HC itself. Backlash is set in the capture software. A hassle. I never touch this HC. I only switch on the 12V and off. The Pegasus should be able to rule out the Lacerta HC, according to Evangelis. Then I will have only the Pegasus UPB sitting on my telescope. Nice and clean.

**Comment by mworion** [comment]
Arie,
in v40 I will have a revised ASCOM handling, which unfortunately does not make the lacerta handling better. The Pegasus UPBv2 focuser should be supported, but I cannot test in reality as I use Sesto Senso v1.
Michel

**Comment by ACMBakx** [comment]
[mw4-2021-08-15.log](https://github.com/mworion/MountWizzard4/files/6988559/mw4-2021-08-15.log)
Quick run with new dimension for the dome opening.
Now it works ok.
Please check in the log why I couldn't slew to Denebola but I could to Arcturus.
So I slewed to Arcturus and the dome moved along.
Then I could slew to Denebola and the dome followed along.

**Comment by mworion** [comment]
v2.0.0b40 published
